# Movie Data ETL Pipeline - Transform Wikipedia Data

### Dependencies

In [1]:
%matplotlib inline

import json
import re
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

### Kaggle data

Download `movies_metadata.csv` and `ratings.csv` from the TMDB's movie dataset at the link below. Move both files into the `data/raw/` directory.

Source: https://www.kaggle.com/rounakbanik/the-movies-dataset

In [2]:
# Path to data directory
data_path = '../data/'

In [3]:
# Kaggle movie metadata
kmovies_df = pd.read_csv(data_path + 'raw/movies_metadata.csv', low_memory=False)
print(kmovies_df.info())
kmovies_df.head(2)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 45466 entries, 0 to 45465
Data columns (total 24 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   adult                  45466 non-null  object 
 1   belongs_to_collection  4494 non-null   object 
 2   budget                 45466 non-null  object 
 3   genres                 45466 non-null  object 
 4   homepage               7782 non-null   object 
 5   id                     45466 non-null  object 
 6   imdb_id                45449 non-null  object 
 7   original_language      45455 non-null  object 
 8   original_title         45466 non-null  object 
 9   overview               44512 non-null  object 
 10  popularity             45461 non-null  object 
 11  poster_path            45080 non-null  object 
 12  production_companies   45463 non-null  object 
 13  production_countries   45463 non-null  object 
 14  release_date           45379 non-null  object 
 15  re

adult                              belongs_to_collection    budget  \
0  False  {'id': 10194, 'name': 'Toy Story Collection', ...  30000000   
1  False                                                NaN  65000000   

                                              genres  \
0  [{'id': 16, 'name': 'Animation'}, {'id': 35, '...   
1  [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...   

                               homepage    id    imdb_id original_language  \
0  http://toystory.disney.com/toy-story   862  tt0114709                en   
1                                   NaN  8844  tt0113497                en   

  original_title                                           overview  ...  \
0      Toy Story  Led by Woody, Andy's toys live happily in his ...  ...   
1        Jumanji  When siblings Judy and Peter discover an encha...  ...   

  release_date      revenue runtime  \
0   1995-10-30  373554033.0    81.0   
1   1995-12-15  262797249.0   104.0   

                                    spoken_languages    status  \
0           [{'iso_639_1': 'en', 'name': 'English'}]  Released   
1  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...  Released   

                                     tagline      title  video vote_average  \
0                                        NaN  Toy Story  False          7.7   
1  Roll the dice and unleash the excitement!    Jumanji  False          6.9   

  vote_count  
0     5415.0  
1     2413.0  

[2 rows x 24 columns]

In [4]:
# Kaggle ratings data
kratings_df = pd.read_csv(data_path + 'raw/ratings.csv')
print(kratings_df.info(null_counts=True))
kratings_df.head(2)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 26024289 entries, 0 to 26024288
Data columns (total 4 columns):
 #   Column     Non-Null Count     Dtype  
---  ------     --------------     -----  
 0   userId     26024289 non-null  int64  
 1   movieId    26024289 non-null  int64  
 2   rating     26024289 non-null  float64
 3   timestamp  26024289 non-null  int64  
dtypes: float64(1), int64(3)
memory usage: 794.2 MB
None


userId  movieId  rating   timestamp
0       1      110     1.0  1425941529
1       1      147     4.5  1425942435

### Wikipedia data

This data was scraped from the movie pages on [Wikipedia](https://en.wikipedia.org/).

In [5]:
# Wikipedia movie data
wmovies_df = pd.read_json(data_path + 'raw/wikipedia.movies.json')
print(wmovies_df.info())
wmovies_df.head(2)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7311 entries, 0 to 7310
Columns: 193 entries, url to Polish
dtypes: float64(5), object(188)
memory usage: 10.8+ MB
None


url    year  \
0  https://en.wikipedia.org/wiki/The_Adventures_o...  1990.0   
1  https://en.wikipedia.org/wiki/After_Dark,_My_S...  1990.0   

                               imdb_link                            title  \
0  https://www.imdb.com/title/tt0098987/  The Adventures of Ford Fairlane   
1  https://www.imdb.com/title/tt0098994/             After Dark, My Sweet   

    Directed by                  Produced by  \
0  Renny Harlin   [Steve Perry, Joel Silver]   
1   James Foley  [Ric Kidney, Robert Redlin]   

                                Screenplay by                     Story by  \
0  [David Arnott, James Cappe, Daniel Waters]  [David Arnott, James Cappe]   
1                [James Foley, Robert Redlin]                          NaN   

                                            Based on  \
0                        [Characters, by Rex Weiner]   
1  [the novel, After Dark, My Sweet, by, Jim Thom...   

                                            Starring  ... Predecessor  \
0  [Andrew Dice Clay, Wayne Newton, Priscilla Pre...  ...         NaN   
1  [Jason Patric, Rachel Ward, Bruce Dern, George...  ...         NaN   

  Founders Area served Products Services Russian Hebrew Revenue  \
0      NaN         NaN      NaN      NaN     NaN    NaN     NaN   
1      NaN         NaN      NaN      NaN     NaN    NaN     NaN   

  Operating income Polish  
0              NaN    NaN  
1              NaN    NaN  

[2 rows x 193 columns]

In [6]:
# Inspect columns
print(wmovies_df.columns.sort_values().tolist())

['Actor control', 'Adaptation by', 'Alias', 'Alma mater', 'Also known as', 'Animation by', 'Arabic', 'Area', 'Area served', 'Artist(s)', 'Attraction type', 'Audio format', 'Author', 'Based on', 'Biographical data', 'Bopomofo', 'Born', 'Box office', 'Budget', 'Camera setup', 'Cantonese', 'Characters', 'Children', 'Chinese', 'Cinematography', 'Closing date', 'Color process', 'Comics', 'Composer(s)', 'Coordinates', 'Country', 'Country of origin', 'Cover artist', 'Created by', 'Date premiered', 'Designer(s)', 'Developed by', 'Developer(s)', 'Dewey Decimal', 'Died', 'Directed by', 'Director', 'Distributed by', 'Distributor', 'Divisions', 'Duration', 'Edited by', 'Editor(s)', 'Ending theme', 'Engine', 'Engine(s)', 'Executive producer(s)', 'Family', 'Fate', 'Film(s)', 'Followed by', 'Format(s)', 'Formerly', 'Founded', 'Founder', 'Founders', 'French', 'Full name', 'Gender', 'Genre', 'Genre(s)', 'Genres', 'Gwoyeu Romatzyh', 'Hangul', 'Hanyu Pinyin', 'Headquarters', 'Hebrew', 'Height', 'Hepburn'

Since the Wikipedia data is a lot messier than the Kaggle data, Reading the it in as a Pandas dataframe resulted in 193 columns and a lot of null values. Instead, we'll be reading this data in as JSON and cleaning it up before converting it to a dataframe.

In [7]:
# Wikipedia movie data
with open(data_path + 'raw/wikipedia.movies.json', 'r') as f:
    wmovies = json.load(f)
    
print('Number of records:', len(wmovies))
print('Sample record:')
wmovies[0]

Number of records: 7311
Sample record:


{'url': 'https://en.wikipedia.org/wiki/The_Adventures_of_Ford_Fairlane',
 'year': 1990,
 'imdb_link': 'https://www.imdb.com/title/tt0098987/',
 'title': 'The Adventures of Ford Fairlane',
 'Directed by': 'Renny Harlin',
 'Produced by': ['Steve Perry', 'Joel Silver'],
 'Screenplay by': ['David Arnott', 'James Cappe', 'Daniel Waters'],
 'Story by': ['David Arnott', 'James Cappe'],
 'Based on': ['Characters', 'by Rex Weiner'],
 'Starring': ['Andrew Dice Clay',
  'Wayne Newton',
  'Priscilla Presley',
  'Lauren Holly',
  'Morris Day',
  'Robert Englund',
  "Ed O'Neill"],
 'Narrated by': 'Andrew "Dice" Clay',
 'Music by': ['Cliff Eidelman', 'Yello'],
 'Cinematography': 'Oliver Wood',
 'Edited by': 'Michael Tronick',
 'Productioncompany ': 'Silver Pictures',
 'Distributed by': '20th Century Fox',
 'Release date': ['July 11, 1990', '(', '1990-07-11', ')'],
 'Running time': '102 minutes',
 'Country': 'United States',
 'Language': 'English',
 'Budget': '$20 million',
 'Box office': '$21.4 milli

### Filter for movies

There seem to be TV shows mixed into the data. To filter for movies, records must have a value for:
1. `imdb_link`
2. `Directed by`/`Director`
3. `Duration`/`Length`/`Running time`

Records should also NOT have a value for:
1. `No. of seasons`
2. `No. of episodes`

In [8]:
# Filter out tv shows
wmovies_filtered = [movie for movie in wmovies if 
                    ('imdb_link' in movie) and 
                    ('Directed by' in movie or 'Director' in movie) and 
                    ('Duration' in movie or 'Length' in movie or 'Running time' in movie) and 
                    ('No. of seasons' not in movie) and 
                    ('No. of episodes' not in movie)]
len(wmovies_filtered)

6936

In [9]:
# Convert to df
wmovies_filtered_df = pd.DataFrame(wmovies_filtered)
print(wmovies_filtered_df.shape)

# Inspect columns
wmovies_cols = wmovies_filtered_df.columns.sort_values()
wmovies_cols

(6936, 75)


Index(['Adaptation by', 'Also known as', 'Animation by', 'Arabic',
       'Audio format', 'Based on', 'Box office', 'Budget', 'Cantonese',
       'Chinese', 'Cinematography', 'Color process', 'Composer(s)', 'Country',
       'Country of origin', 'Created by', 'Directed by', 'Director',
       'Distributed by', 'Distributor', 'Edited by', 'Editor(s)',
       'Executive producer(s)', 'Followed by', 'French', 'Genre', 'Hangul',
       'Hebrew', 'Hepburn', 'Japanese', 'Label', 'Language', 'Length',
       'Literally', 'Mandarin', 'McCune–Reischauer', 'Music by', 'Narrated by',
       'Original language(s)', 'Original network', 'Original release',
       'Original title', 'Picture format', 'Polish', 'Preceded by',
       'Produced by', 'Producer', 'Producer(s)', 'Production company(s)',
       'Production location(s)', 'Productioncompanies ', 'Productioncompany ',
       'Recorded', 'Release date', 'Released', 'Revised Romanization',
       'Romanized', 'Running time', 'Russian', 'Screen st

### Clean columns

Just by filtering out the TV shows, the data has been reduced to 75 columns. Let's take a look at a sample value for each column to get a better understanding of the data. Since there are a lot of missing values in most of the columns, the sample values will not all correspond to the same record.

In [10]:
# Sample value for each column
for col in wmovies_cols:
    # Print the first non-null value in the column
    print(col, ':', wmovies_filtered_df.loc[wmovies_filtered_df[col].notnull(), col].values[0])

Adaptation by : ['John L. Balderston', 'Paul Perez', 'Daniel Moore']
Also known as : Detonator II: Night Watch
Animation by : ['Andreas Deja', 'Gary Dunn', 'Deboissy Sylvain']
Arabic : قضية رقم ٢٣
Audio format : Stereo
Based on : ['Characters', 'by Rex Weiner']
Box office : $21.4 million
Budget : $20 million
Cantonese : ['Jip', '6', 'Man', '6', 'Saam', '1']
Chinese : 摇滚藏獒
Cinematography : Oliver Wood
Color process : Technicolor
Composer(s) : Richard Bellis
Country : United States
Country of origin : United States
Created by : ['John William Corrington', '(novel)']
Directed by : Renny Harlin
Director : Mark "Aldo" Miceli
Distributed by : 20th Century Fox
Distributor : NBC
Edited by : Michael Tronick
Editor(s) : ['Christopher Cooke', 'James Galloway']
Executive producer(s) : Rich Melcombe
Followed by : See below
French : Le Cinquième Élément
Genre : Thriller
Hangul : 원더풀 데이즈
Hebrew : פוֹקְסטְרוֹט
Hepburn : Omoide no Mānī
Japanese : 思い出のマーニー
Label : ['Warner Music Vision', 'Warner-Reprise

We will be addressing a few things here:
1. There are a lot of columns holding alternate titles for the movies so we're going to group all of these together in the JSON data.
2. There are also a lot of redundant columns giving the same information, such as `Directed by` and `Director`. We'll be grouping these together as well.
3. The column names are inconsistent, so we will rename them for consistency.

In [11]:
# Keys holding alternate titles
title_keys = ['Also known as', 'Arabic', 'Cantonese', 'Chinese', 'French', 
              'Hangul', 'Hebrew', 'Hepburn', 'Japanese', 'Literally', 'Mandarin', 
              'McCune–Reischauer', 'Original title', 'Polish', 'Revised Romanization', 
              'Romanized', 'Russian', 'Simplified', 'Traditional', 'Yiddish']

# Key rename pairs (old name: new name)
keys_to_rename = {
    'Adaptation by': 'writers',
    'Animation by': 'animators',
    'Audio format': 'audio_format',
    'Based on': 'based_on',
    'Box office': 'box_office',
    'Budget': 'budget',
    'Cinematography': 'cinematographers',
    'Color process': 'color_process',
    'Composer(s)': 'composers',
    'Country': 'country',
    'Country of origin': 'country',
    'Created by': 'creators',
    'Directed by': 'director',
    'Director': 'director',
    'Distributed by': 'distributor',
    'Distributor': 'distributor',
    'Edited by': 'editors',
    'Editor(s)': 'editors',
    'Executive producer(s)': 'executive_producers',
    'Followed by': 'sequel',
    'Genre': 'genre',
    'Label': 'label',
    'Language': 'languages',
    'Length': 'duration',
    'Music by': 'composers',
    'Narrated by': 'narrator',
    'Original language(s)': 'languages', 
    'Original network': 'network', 
    'Original release': 'release_date',
    'Picture format': 'picture_format',
    'Preceded by': 'prequel',
    'Produced by': 'producers',
    'Producer': 'producers',
    'Producer(s)': 'producers',
    'Production company(s)': 'production_companies',
    'Production location(s)': 'production_locations',
    'Productioncompanies ': 'production_companies',
    'Productioncompany ': 'production_companies',
    'Recorded': 'recorded',
    'Release date': 'release_date',
    'Released': 'release_date',
    'Running time': 'duration',
    'Screen story by': 'writers',
    'Screenplay by': 'writers',
    'Starring': 'stars',
    'Story by': 'writers',
    'Suggested by': 'suggestors',
    'Theme music composer': 'composers',
    'Venue': 'venue',
    'Voices of': 'voicers',
    'Written by': 'writers',
    'imdb_link': 'imdb_link', 
    'title': 'title',
    'url': 'url', 
    'year': 'year'
}

len(title_keys), len(keys_to_rename)

(20, 55)

In [12]:
def clean_movie(movie_dict, title_keys=title_keys, keys_to_rename=keys_to_rename):
    
    """
    Clean movie dictionary with the following steps:
        [1] combine all alternate titles into a single key
        [2] rename keys for consistency and to consolidate similar columns into 1
    
    Args:
        [1] movie_dict (dict) - record to clean
        [2] title_keys (list[str]) - names of keys with the movie's alternate titles
        [3] keys_to_rename (dict{str: str}) - mapping of old key name to new key name
    
    Returns:
        (Dict) clean movie dictionary
    """
    
    # Copy of movie dict and empty dict for alternate titles
    movie_dict, alt_titles = dict(movie_dict), dict()
    
    # Add keys with alternate titles to alt_titles dict and delete the original key
    for key in title_keys:
        if key in movie_dict:
            alt_titles[key.lower().replace(' ', '_')] = movie_dict.pop(key)
            
    # Add new key for alternate titles
    if len(alt_titles):
        movie_dict['alternate_titles'] = alt_titles
        
    # Rename keys
    for old, new in keys_to_rename.items():
        if old in movie_dict:
            movie_dict[new] = movie_dict.pop(old)
        
    return movie_dict
            
    
# Test function
clean_movie(wmovies_filtered[849])

{'alternate_titles': {'mandarin': 'Xǐyàn', 'traditional': '喜宴'},
 'box_office': '$23.6 million',
 'budget': '$1 million',
 'cinematographers': 'Jong Lin',
 'country': ['Taiwan', 'United States'],
 'director': 'Ang Lee',
 'distributor': 'The Samuel Goldwyn Company',
 'editors': 'Tim Squyres',
 'languages': ['Mandarin Chinese', 'English'],
 'composers': 'Mader',
 'producers': ['Ang Lee', 'Ted Hope', 'James Schamus'],
 'production_companies': 'Good Machine',
 'release_date': ['4 August 1993', '(', '1993-08-04', ')', '(United States)'],
 'duration': '106 minutes',
 'stars': ['Ah-Leh Gua',
  'Sihung Lung',
  'May Chin',
  'Winston Chao',
  'Mitchell Lichtenstein'],
 'writers': ['Ang Lee', 'Neil Peng', 'James Schamus'],
 'imdb_link': 'https://www.imdb.com/title/tt0107156/',
 'title': 'The Wedding Banquet',
 'url': 'https://en.wikipedia.org/wiki/The_Wedding_Banquet',
 'year': 1993}

In [13]:
# Clean movie dictionaries
wmovies_clean = [clean_movie(movie) for movie in wmovies_filtered]

# Convert Wikipedia data to df
wmovies_clean_df = pd.DataFrame(wmovies_clean)
print(wmovies_clean_df.shape)

# Inspect columns
wmovies_clean_df.columns.sort_values()

(6936, 38)


Index(['alternate_titles', 'animators', 'audio_format', 'based_on',
       'box_office', 'budget', 'cinematographers', 'color_process',
       'composers', 'country', 'creators', 'director', 'distributor',
       'duration', 'editors', 'executive_producers', 'genre', 'imdb_link',
       'label', 'languages', 'narrator', 'network', 'picture_format',
       'prequel', 'producers', 'production_companies', 'production_locations',
       'recorded', 'release_date', 'sequel', 'stars', 'suggestors', 'title',
       'url', 'venue', 'voicers', 'writers', 'year'],
      dtype='object')

### Drop duplicate rows

With the columns taken care of, let's inspect the data for duplicate records. The IMDB ID should be unique for each movie. The data does have the IMDB link, so we can extract the IMDB ID from that link.

In [14]:
# Inspect IMDB links
wmovies_clean_df.loc[:1, 'imdb_link']

0    https://www.imdb.com/title/tt0098987/
1    https://www.imdb.com/title/tt0098994/
Name: imdb_link, dtype: object

In [15]:
# Extract IMDB ID from the link
# wmovies_clean_df['id'] = wmovies_clean_df['imdb_link'].str.split('/', expand=True)[4]
wmovies_clean_df['id'] = wmovies_clean_df['imdb_link'].str.extract(r'(tt\d{7})')
wmovies_clean_df.head(2)

based_on     box_office  \
0                        [Characters, by Rex Weiner]  $21.4 million   
1  [the novel, After Dark, My Sweet, by, Jim Thom...   $2.7 million   

        budget cinematographers        country      director  \
0  $20 million      Oliver Wood  United States  Renny Harlin   
1   $6 million     Mark Plummer  United States   James Foley   

        distributor          editors languages                composers  ...  \
0  20th Century Fox  Michael Tronick   English  [Cliff Eidelman, Yello]  ...   
1   Avenue Pictures  Howard E. Smith   English            Maurice Jarre  ...   

  creators prequel suggestors alternate_titles label recorded venue animators  \
0      NaN     NaN        NaN              NaN   NaN      NaN   NaN       NaN   
1      NaN     NaN        NaN              NaN   NaN      NaN   NaN       NaN   

  color_process         id  
0           NaN  tt0098987  
1           NaN  tt0098994  

[2 rows x 39 columns]

In [16]:
# Check for duplicate movies
print(wmovies_clean_df.duplicated(subset=['id']).sum())
wmovies_clean_df[wmovies_clean_df.duplicated(subset=['id'], keep=False)].sort_values('id').head(4)

42


based_on     box_office  \
23   [Characters, by, H. P. Lovecraft]            NaN   
273  [Characters, by, H. P. Lovecraft]            NaN   
199                                NaN  $15.1 million   
611                                NaN  $15.1 million   

                             budget          cinematographers  \
23                              NaN              Rick Fichter   
273                             NaN              Rick Fichter   
199  [60 million, Norwegian Kroner]  Erling Thurmann-Andersen   
611  [60 million, Norwegian Kroner]  Erling Thurmann-Andersen   

                             country     director           distributor  \
23                     United States  Brian Yuzna     50th Street Films   
273                    United States  Brian Yuzna     50th Street Films   
199  [Norway, Sweden, United States]    Nils Gaup  Buena Vista Pictures   
611  [Norway, Sweden, United States]    Nils Gaup  Buena Vista Pictures   

                editors languages      composers  ... creators prequel  \
23       Peter Teschner   English   Richard Band  ...      NaN     NaN   
273      Peter Teschner   English   Richard Band  ...      NaN     NaN   
199  Nils Pagh Andersen   English  Patrick Doyle  ...      NaN     NaN   
611  Nils Pagh Andersen   English  Patrick Doyle  ...      NaN     NaN   

    suggestors alternate_titles label recorded venue animators color_process  \
23         NaN              NaN   NaN      NaN   NaN       NaN           NaN   
273        NaN              NaN   NaN      NaN   NaN       NaN           NaN   
199        NaN              NaN   NaN      NaN   NaN       NaN           NaN   
611        NaN              NaN   NaN      NaN   NaN       NaN           NaN   

            id  
23   tt0099180  
273  tt0099180  
199  tt0099816  
611  tt0099816  

[4 rows x 39 columns]

In [17]:
# Drop duplicate movies
wmovies_clean_df.drop_duplicates(subset=['id'], inplace=True)
wmovies_clean_df.shape

(6894, 39)

### Drop columns with mostly missing values

There are a lot of columns that are missing most of their values, so there is no point in keeping them. We will be looking for and dropping columns with at least half of their values missing.

In [18]:
# Check columns where at least half of the values are missing
cols_missing = wmovies_clean_df.isnull().mean()
cols_missing70 = cols_missing[cols_missing >= 0.5]
cols_missing70

based_on                0.684798
narrator                0.959240
genre                   0.984624
network                 0.982594
audio_format            0.991297
executive_producers     0.986365
picture_format          0.991007
production_locations    0.993328
sequel                  0.998695
voicers                 0.999710
creators                0.998549
prequel                 0.998549
suggestors              0.999855
alternate_titles        0.996954
label                   0.999710
recorded                0.999710
venue                   0.999855
animators               0.999710
color_process           0.999855
dtype: float64

In [19]:
# Drop these columns and inspect remaining columns
wmovies_clean_df.drop(cols_missing70.index, axis=1, inplace=True)
print(wmovies_clean_df.shape)
wmovies_clean_df.isnull().mean()

(6894, 20)


box_office              0.208297
budget                  0.315202
cinematographers        0.093560
country                 0.031912
director                0.000000
distributor             0.043951
editors                 0.070351
languages               0.010299
composers               0.067740
producers               0.025239
production_companies    0.231651
release_date            0.003626
duration                0.000000
writers                 0.026690
stars                   0.023934
imdb_link               0.000000
title                   0.000000
url                     0.000000
year                    0.000000
id                      0.000000
dtype: float64

### Inspect data types

With 20 columns remaining, let's check if their data types are appropriate.

In [20]:
# Inspect columns
print(wmovies_clean_df.info())
wmovies_clean_df.head(2)

<class 'pandas.core.frame.DataFrame'>
Int64Index: 6894 entries, 0 to 6935
Data columns (total 20 columns):
 #   Column                Non-Null Count  Dtype 
---  ------                --------------  ----- 
 0   box_office            5458 non-null   object
 1   budget                4721 non-null   object
 2   cinematographers      6249 non-null   object
 3   country               6674 non-null   object
 4   director              6894 non-null   object
 5   distributor           6591 non-null   object
 6   editors               6409 non-null   object
 7   languages             6823 non-null   object
 8   composers             6427 non-null   object
 9   producers             6720 non-null   object
 10  production_companies  5297 non-null   object
 11  release_date          6869 non-null   object
 12  duration              6894 non-null   object
 13  writers               6710 non-null   object
 14  stars                 6729 non-null   object
 15  imdb_link             6894 non-null   

box_office       budget cinematographers        country      director  \
0  $21.4 million  $20 million      Oliver Wood  United States  Renny Harlin   
1   $2.7 million   $6 million     Mark Plummer  United States   James Foley   

        distributor          editors languages                composers  \
0  20th Century Fox  Michael Tronick   English  [Cliff Eidelman, Yello]   
1   Avenue Pictures  Howard E. Smith   English            Maurice Jarre   

                     producers production_companies  \
0   [Steve Perry, Joel Silver]      Silver Pictures   
1  [Ric Kidney, Robert Redlin]      Avenue Pictures   

                                        release_date     duration  \
0                  [July 11, 1990, (, 1990-07-11, )]  102 minutes   
1  [May 17, 1990, (, 1990-05-17, ), (Cannes Film ...  114 minutes   

                        writers  \
0   [David Arnott, James Cappe]   
1  [James Foley, Robert Redlin]   

                                               stars  \
0  [Andrew Dice Clay, Wayne Newton, Priscilla Pre...   
1  [Jason Patric, Rachel Ward, Bruce Dern, George...   

                               imdb_link                            title  \
0  https://www.imdb.com/title/tt0098987/  The Adventures of Ford Fairlane   
1  https://www.imdb.com/title/tt0098994/             After Dark, My Sweet   

                                                 url  year         id  
0  https://en.wikipedia.org/wiki/The_Adventures_o...  1990  tt0098987  
1  https://en.wikipedia.org/wiki/After_Dark,_My_S...  1990  tt0098994

Columns that need to be recasted:
- `Release_date` - to datetime
- `Budget` - to numeric
- `Box_office` - to numeric
- `Duration` - to numeric

### Convert `release_date` to datetime type

The data contains lists, which are unhashable, so we'll need to join the list items into a single string before making the conversion.

In [27]:
def list_to_str(obj):
    
    """ Convert string to a list """
    
    return ' '.join(obj) if isinstance(obj, list) else obj


# Convert lists in release dates to strings
release_date = wmovies_clean_df['release_date'].apply(list_to_str)
release_date.unique().tolist()

['July 11, 1990 ( 1990-07-11 )',
 'May 17, 1990 ( 1990-05-17 ) (Cannes Film Market) August 24, 1990 ( 1990-08-24 ) (United States)',
 'August 10, 1990 ( 1990-08-10 )',
 'December 25, 1990 ( 1990-12-25 )',
 'December 19, 1990',
 'March 22, 1990 ( 1990-03-22 )',
 'October 6, 1990 ( 1990-10-06 ) ( New York Film Festival )',
 'March 8, 1991 ( 1991-03-08 )',
 'March 7, 1990 ( 1990-03-07 )',
 'February 23, 1990 ( USA )',
 'June 8, 1990 ( 1990-06-08 )',
 'July 18, 1990 ( 1990-07-18 )',
 'October 5, 1990 ( 1990-10-05 ) (United States)',
 'December 20, 1990 (limited U.S.) January 11, 1991 (wide U.S.)',
 'May 25, 1990 ( 1990-05-25 )',
 'March 9, 1990',
 'March 2, 1990 ( 1990-03-02 )',
 'June 22, 1990 ( 1990-06-22 )',
 'May 18, 1990 ( 1990-05-18 )',
 'February 11, 1990',
 'October 12, 1990 ( 1990-10-12 )',
 'January 1990 ( 1990-01 ) ( Sundance Film Festival ) March 16, 1990 ( 1990-03-16 )',
 'December 21, 1990 ( 1990-12-21 )',
 'September 8, 1990 ( 1990-09-08 ) ( TIFF ) February 22, 1991 ( 1991-0

Some of the dates in the data are actually ranges (ex. January 15 - 16, 1990). We will be removing the right limit of these date ranges and keeping the left limit. In other words, `January 15 - 16, 1990` would be transformed into `January 15, 1990`. After addressing this issue, there are 5 different formats the dates were recorded in:
1. (DD) (Month) (YYYY) - 01 January 1970
2. (Month) (DD), (YYYY) - January 01, 1970
3. (Month), (YYYY) - January, 1970
4. (YYYY)-(MM)-(DD) - 1970-01-01
5. (YYYY) - 1970

These formats are all parsable by Pandas, so we will be extracting these patterns from the date strings in order to convert them into datetime type.

In [22]:
# def standardize_date(d):
    
#     """ Standardize a date string into the format YYYY-MM-DD """
    
#     months = {
#         'January': '01',
#         'February': '02',
#         'March': '03',
#         'April': '04',
#         'May': '05',
#         'June': '06',
#         'July': '07',
#         'August': '08',
#         'September': '09',
#         'October': '10',
#         'November': '11',
#         'December': '12'
#     }
    
#     # Initialize date variables
#     YYYY = MM = DD = None
    
#     # For format: (Month) (DD), (YYYY)
#     match = re.search(r'([JFMASOND][a-z]{2,8}) (\d\d?), (\d{4})', d)
#     if match:
#         DD = match.group(2)
#         DD = '0' + DD if len(DD) == 1 else DD
#         MM = months[match.group(1)]
#         YYYY = match.group(3)

#     # For format: (DD) (Month) (YYYY)
#     match = re.search(r'(\d\d?) ([JFMASOND][a-z]{2,8}) (\d{4})', d)
#     if match:
#         DD = match.group(1)
#         DD = '0' + DD if len(DD) == 1 else DD
#         MM = months[match.group(2)]
#         YYYY = match.group(3)
        
#     # For format: (Month),? (YYYY)
#     match = re.search(r'([JFMASOND][a-z]{2,8}),? (\d{4})', d)
#     if match:
#         MM = months[match.group(1)]
#         YYYY = match.group(2)
       
#     # Standard format: YYYY-MM-DD
#     date = d if not YYYY else YYYY + '-' + MM 
#     return date if not DD else date + '-' + DD
    

# def extract_date(s):
    
#     """
#     Extract date from string
#     Note: In a range, select the left limit
#         Ex. December 15 - 20, 1999 becomes December 15, 1999
#         Ex. 1990 - 1992 becomes 1990
#     """
    
#     # Null check
#     if pd.isnull(s):
#         return s
    
#     # Select left limit of ranges
#     s = re.sub(r' [-–—] \d\d?', '', s.strip()) # include hyphen, en dash, em dash
    
#     # For format: (DD)? (Month) (DD)?,? (YYYY)
#     formats = [r'(?:\d\d? )?[JFMASOND][a-z]{2,8}(?: \d\d?)?,? \d{4}']
    
#     # For format: (YYYY)-(MM)-(DD)
#     formats.append(r'\d{4}(?:\D\d\d?\D\d\d?)?')
    
#     for f in formats:
#         match = re.search(f, s)
#         if match:
#             return standardize_date(match.group(0))


# # Test functions
# dates_extracted, dates_not_extracted = [], []
# for s in release_date.tolist():
#     d = extract_date(s)
#     if d is None:
#         dates_not_extracted.append(s)
#     else:
#         dates_extracted.append(d)
# len(dates_extracted), len(dates_not_extracted)

In [23]:
# # Convert to datetime
# release_date = pd.to_datetime(release_date.apply(extract_date))
# release_date

In [28]:
# Select lower limit of date ranges
release_date = release_date.str.strip().str.replace(r' [-–—] \d\d?', '', regex=True)

In [29]:
# Date formats
date_format1 = r'(?:\d\d? )?[a-z]{3,9}(?: \d\d?)?,? \d{4}' # (DD) (Month) (YYYY) | (Month) (DD), (YYYY) | (Month), (YYYY)
date_format2 = r'\d{4}(?:\D\d\d?\D\d\d?)?' # (YYYY)-(MM)-(DD) | (YYYY)
date_formats = f'({date_format1}|{date_format2})' # date formats 1 and 2

# Check if there are any other dates not captured by these 2 patterns
release_date.dropna()[~release_date.dropna().str.contains(date_formats, flags=re.IGNORECASE)]

/Users/tribui/miniconda3/envs/minds/lib/python3.7/site-packages/pandas/core/strings.py:2001: UserWarning: This pattern has match groups. To actually get the groups, use str.extract.
  return func(self, *args, **kwargs)


Series([], Name: release_date, dtype: object)

In [30]:
# Extract date from string
release_date = release_date.str.extract(date_formats, flags=re.IGNORECASE)[0]

# Convert date string to datetime type
wmovies_clean_df['release_date'] = pd.to_datetime(release_date, infer_datetime_format=True)
wmovies_clean_df['release_date'].isnull().sum()

25

### Convert `budget` to numeric

As with `release_date`, there are lists and ranges in the `budget` values, so we will convert any list to strings and reduce ranges to the lower limit the same way. In addition, `budget` also has citation brackets (ex. [1]) which we will removing as well.

In [34]:
# Convert lists in `budget` to strings
budget = wmovies_clean_df['budget'].apply(list_to_str)

# Clean string and select lower limit of amount ranges
budget = budget.str.strip().str.replace(r'\[\s*(?:\w+\s*)*\]', '', regex=True) \
                           .str.replace(r'[-–—]\s?\$?\d+', '', regex=True)
budget.unique().tolist()

['$20 million',
 '$6 million',
 '$35 million',
 '$12 million',
 '$25 million',
 nan,
 '$50 million',
 '$22 million',
 '$29 million',
 '$40 million',
 '$7 million',
 '$2,500,000 ',
 '$47 million',
 '$15 million',
 '$10 million',
 '$13 million',
 '$5.2 million',
 '$17.5 million',
 'US$18 million',
 '$4 million',
 '$16 million',
 '$60 million',
 '$6.5 million',
 '$18 million',
 '$46 million',
 '$70 million',
 '$10,000,000 (estimated)',
 '$19,000,000',
 '$11 million',
 '$500,000',
 '$10 million (estimated)',
 '$26 million',
 '$11.5 million',
 '$54 million',
 '$12.5 million',
 '$3 million',
 '$13,000,000',
 '£960,000  ($1.5 million)',
 '$110,000',
 '$2.5 million',
 '$30 million',
 '$5 million',
 '$9 million',
 'USD$ 9 million',
 'Unknown',
 '$23 million (estimated) ',
 '$225,000',
 '$14 million ',
 '$21 million',
 '$4.2 million',
 '$2,000,000 (estimated)',
 '$14 million',
 '$6,000,000',
 '$17 million',
 '$1.5 million',
 '$20 million ',
 '$6.5 million  ',
 '$42 million',
 '60 million Norwegi

It looks like 2 formats will capture most or all of the `budget` values:
1. \\$xxx.x mil - \\$100 million, \\$1.2 million
2. \\$xxx,xxx,xxx - \\$100,000,000, \\$20,000

In [39]:
# Budget formats
budget_format1 = r'\$?\s?\d{1,3}(?:\.\d+)?\s*mil' # $xxx.x mil
budget_format2 = r'\$?\s?\d{1,3}(?:,\d{3})+' # $xxx,xxx,xxx
budget_formats = f'({budget_format1}|{budget_format2})'

# Check formats
budget_contains = budget.dropna().str.contains(budget_formats, flags=re.IGNORECASE)
budget_to_na = budget.dropna()[~budget_contains].unique().tolist()
budget_to_na

['Unknown', 'HBO', '$218.32', 'N/A', '19 crore', '3.5 crore']

In [40]:
# Replace above values with NaN
for val in budget_to_na:
    budget.replace(val, np.NaN, inplace=True)
budget.dropna()[~budget_contains]

Series([], Name: budget, dtype: object)

In [41]:
# Extract amount from string
budget = budget.str.extract(budget_formats, flags=re.IGNORECASE)[0]
# budget.unique().tolist()

['$20 mil',
 '$6 mil',
 '$35 mil',
 '$12 mil',
 '$25 mil',
 nan,
 '$50 mil',
 '$22 mil',
 '$29 mil',
 '$40 mil',
 '$7 mil',
 '$2,500,000',
 '$47 mil',
 '$15 mil',
 '$10 mil',
 '$13 mil',
 '$5.2 mil',
 '$17.5 mil',
 '$18 mil',
 '$4 mil',
 '$16 mil',
 '$60 mil',
 '$6.5 mil',
 '$46 mil',
 '$70 mil',
 '$10,000,000',
 '$19,000,000',
 '$11 mil',
 '$500,000',
 '$26 mil',
 '$11.5 mil',
 '$54 mil',
 '$12.5 mil',
 '$3 mil',
 '$13,000,000',
 '960,000',
 '$110,000',
 '$2.5 mil',
 '$30 mil',
 '$5 mil',
 '$9 mil',
 '$ 9 mil',
 '$23 mil',
 '$225,000',
 '$14 mil',
 '$21 mil',
 '$4.2 mil',
 '$2,000,000',
 '$6,000,000',
 '$17 mil',
 '$1.5 mil',
 '$42 mil',
 '60 mil',
 '$350,000',
 '$43 mil',
 '$250,000',
 '$19 mil',
 '$3.5 mil',
 '$13.5 mil',
 '$24 mil',
 '$ 80,000',
 '$36 mil',
 '$48 mil',
 '$800,000',
 '$38 mil',
 '$15,000,000',
 '$6.788 mil',
 '$ 65 mil',
 '$75 mil',
 '$4.5 mil',
 '$ 5 mil',
 '$8 mil',
 '$ 24 mil',
 '$780,000',
 '$7.5 mil',
 '$11,000,000',
 '$23,000',
 '$27 mil',
 '$450,720',
 '$94 m

With the patterns extracted, we can now parse these values in order to convert them to numeric type.

In [43]:
def parse_budget(s):
    
    """ Convert budget string to float """
    
    # Check if the input is NaN or already a float
    if isinstance(s, float):
        return s
    
    # Remove $, spaces, and commas
    s = re.sub(r'[\$\s,]', '', s).lower()
    
    # Convert to float
    if 'mil' in s:
        f = float(s.replace('mil', '')) * 1e6
    else:
        f = float(s)
        
    return f


# Convert budget to float type
wmovies_clean_df['budget'] = budget.apply(parse_budget)
wmovies_clean_df['budget'].isnull().sum()

2182

### Convert box office to numeric

This will be pretty similar to `budget`, except there are no ranges to worry about. The formats will also be very similar except for the fact that instead of just "million", there are also values in thousands ("k") and "billion"s.

In [48]:
# Convert lists in box office to strings
boxoffice = wmovies_clean_df['box_office'].apply(list_to_str)
# boxoffice.unique().tolist()

In [50]:
# Box office formats
boxoffice_format1 = r'\$?\s?\d{1,3}(?:\.\d+)?\s*[kmb]' # $xxx.x k/m/b
boxoffice_format2 = r'\$?\s?\d{1,3}(?:[\s\.,]?\d{3})+\$?' # $xxx,xxx,xxx
boxoffice_formats = f'({boxoffice_format1}|{boxoffice_format2})'

# Check formats
boxoffice_contains = boxoffice.dropna().str.contains(boxoffice_formats, flags=re.IGNORECASE)
boxoffice_to_na = boxoffice.dropna()[~boxoffice_contains].unique().tolist()
boxoffice_to_na

['N/A',
 '$309',
 'TBA',
 '$20-30',
 '£2.56',
 'Unknown',
 '$588',
 'less than $372',
 '8 crore']

In [51]:
# Replace above values with NaN
for val in boxoffice_to_na:
    boxoffice.replace(val, np.NaN, inplace=True)
boxoffice.dropna()[~boxoffice_contains]

Series([], Name: box_office, dtype: object)

In [52]:
# Extract amount from string
boxoffice = boxoffice.str.extract(boxoffice_formats, flags=re.IGNORECASE)[0]
# boxoffice.unique().tolist()

['$21.4 m',
 '$2.7 m',
 '$57,718,089',
 '$7,331,647',
 '$6,939,946',
 nan,
 '$855,810',
 '$195.3 m',
 '$53.2 m',
 '$15.7 m',
 '$52.1 m',
 '$244.5 m',
 '$12,626,043',
 '$19,740,070',
 '$138,697,012',
 '$1,292,323',
 '$8.2 m',
 '$15.6 m',
 '$2,070,871',
 '$27.6 m',
 '$10,173',
 '$5 m',
 '$35.8 m',
 '$ 4,212,828',
 '$2.4 m',
 '$947,306',
 '$715,983',
 '$13,236,513',
 '$8.3 m',
 '$1,373,728',
 '$424.2 m',
 '$48.8 m',
 '$157.9 m',
 '$2.2 m',
 '$6,698,361',
 '$162.7 m',
 '$240 m',
 '$2,346,150',
 '$18.1 m',
 '$86 m',
 '$25 m',
 '$1,372,350',
 '$39 m',
 '$14,760,451',
 '$22,424,195',
 '$6,488,144',
 '$61.3 m',
 '$1,823,154',
 '$1,305,887',
 '$205,068',
 '$335.000',
 '$21,460,601',
 '$8.1 m',
 '$505.7 m',
 '$136.8 m',
 '$46.8 m',
 '$4.6 m',
 '$29.9 m',
 '$41.5 m',
 '$13,446,769',
 '$17 m',
 '$4,960,385',
 '$47.4 m',
 '$5,728,953',
 '$9,243,140',
 '$4,134,992',
 '$23.5 m',
 '$609,939',
 '$476.7 m',
 '$21.6 m',
 '$1,293,976',
 '$26.4 m',
 '$200.5 m',
 '4.37 m',
 '$16.2 m',
 '$27.7 m',
 '$26.1 m'

In [39]:
def parse_boxoffice(s):
    
    ''' Convert box office string to float '''
    
    # Null check
    if isinstance(s, float):
        return s
    
    # Remove $, spaces, and commas
    s = re.sub(r'[\$\s,]', '', s).lower()
    
    # Convert to float
    if 'k' in s:
        f = float(s.replace('k', '')) * 1e3
    elif 'm' in s:
        f = float(s.replace('m', '')) * 1e6
    elif 'b' in s:
        f = float(s.replace('b', '')) * 1e9
    elif '.' in s:
        f = float(s.replace('.', ''))
    else:
        f = float(s)
        
    return f


# Convert budget to float type
boxoffice_to_float = boxoffice_clean_str.apply(parse_boxoffice)
boxoffice_to_float.isnull().sum()

1445

In [40]:
# Replace release date column
wiki_recast_boxoffice_df = wiki_recast_budget_df.copy()
wiki_recast_boxoffice_df['Box office'] = boxoffice_to_float
wiki_recast_boxoffice_df['Box office']

0       21400000.0
1        2700000.0
2       57718089.0
3        7331647.0
4        6939946.0
           ...    
6931    41900000.0
6932    76100000.0
6933    38400000.0
6934     5500000.0
6935           NaN
Name: Box office, Length: 6894, dtype: float64

### Convert duration to numeric

In [41]:
# Convert lists in duration to strings
duration_to_str = wiki_recast_budget_df['Duration'].apply(list_to_str)

# Clean string and select lower limit of amount ranges
duration_replace_range = duration_to_str.str.strip().str.replace(r'\[\d\]', '', regex=True) \
                                                    .str.replace(r'[-–—]\s?\d+', '', regex=True)
# duration_replace_range.unique().tolist()

In [42]:
# Duration formats
duration_format1 = r'(?:\d\s*[Hh]o?u?r?s?\s*)?\d{1,3}\s*[Mm]'
duration_format2 = r'\d\s*[Hh]o?u?r?s?|\d{1,2}\s*\:\s*\d{1,2}'

# Check formats
duration_contains1 = duration_replace_range.dropna().str.contains(duration_format1, flags=re.IGNORECASE)
duration_contains2 = duration_replace_range.dropna().str.contains(duration_format2, flags=re.IGNORECASE)
duration_to_na = duration_replace_range.dropna()[~duration_contains1 & ~duration_contains2].unique().tolist()
duration_to_na

['varies', 'minutes']

In [43]:
# Replace above values with NaN
duration_replace_vals = duration_replace_range.copy()
for val in duration_to_na:
    duration_replace_vals.replace(val, np.NaN, inplace=True)
duration_replace_vals.dropna()[~duration_contains1 & ~duration_contains2]

Series([], Name: Duration, dtype: object)

In [44]:
# Extract duration from string
duration_clean_str = duration_replace_vals.str.extract(f'({duration_format1}|{duration_format2})', 
                                                       flags=re.IGNORECASE)[0]
# duration_clean_str.unique().tolist()

In [45]:
def parse_duration(s):
    
    ''' Convert duration string to integer '''
    
    # Null check
    if isinstance(s, float):
        return s
    
    # Remove seconds
    s = re.sub(r'\:\s*\d{1,2}', '', s)
    
    # Remove m and spaces
    s = re.sub(r'm|\s*', '', s, flags=re.IGNORECASE)
    
    # Convert to int
    match = re.search(r'(\d)(ho?u?r?s?)(\d\d?)?', s, flags=re.IGNORECASE)
    if match: # if time is in hours
        i = int(match.group(1)) * 60
        if match.group(3):
            i += int(match.group(3))
    else: # if time is in minutes
        i = int(s)
        
    return i


# Convert duration to integer type
duration_to_int = duration_clean_str.apply(parse_duration)
duration_to_int.isnull().sum()

2

In [46]:
# Replace release date column
wiki_recast_duration_df = wiki_recast_boxoffice_df.copy()
wiki_recast_duration_df['Duration'] = duration_to_int
print(wiki_recast_duration_df.info())
wiki_recast_duration_df.head(2)

<class 'pandas.core.frame.DataFrame'>
Int64Index: 6894 entries, 0 to 6935
Data columns (total 21 columns):
 #   Column                 Non-Null Count  Dtype         
---  ------                 --------------  -----         
 0   Based on               2173 non-null   object        
 1   Release date           6867 non-null   datetime64[ns]
 2   Country                6674 non-null   object        
 3   Budget                 4712 non-null   float64       
 4   Box office             5449 non-null   float64       
 5   Cinematographer(s)     6249 non-null   object        
 6   Director               6894 non-null   object        
 7   Distributor            6591 non-null   object        
 8   Editor(s)              6409 non-null   object        
 9   Language(s)            6823 non-null   object        
 10  Composer(s)            6427 non-null   object        
 11  Producer(s)            6718 non-null   object        
 12  Production company(s)  5297 non-null   object        
 13  Dur

Based on Release date  \
0                        [Characters, by Rex Weiner]   1990-07-11   
1  [the novel, After Dark, My Sweet, by, Jim Thom...   1990-05-17   

         Country      Budget  Box office Cinematographer(s)      Director  \
0  United States  20000000.0  21400000.0        Oliver Wood  Renny Harlin   
1  United States   6000000.0   2700000.0       Mark Plummer   James Foley   

        Distributor        Editor(s) Language(s)  ...  \
0  20th Century Fox  Michael Tronick     English  ...   
1   Avenue Pictures  Howard E. Smith     English  ...   

                   Producer(s) Production company(s) Duration  \
0   [Steve Perry, Joel Silver]       Silver Pictures    102.0   
1  [Ric Kidney, Robert Redlin]       Avenue Pictures    114.0   

                      Writer(s)  \
0   [David Arnott, James Cappe]   
1  [James Foley, Robert Redlin]   

                                               Stars  \
0  [Andrew Dice Clay, Wayne Newton, Priscilla Pre...   
1  [Jason Patric, Rachel Ward, Bruce Dern, George...   

                               IMDB link                            Title  \
0  https://www.imdb.com/title/tt0098987/  The Adventures of Ford Fairlane   
1  https://www.imdb.com/title/tt0098994/             After Dark, My Sweet   

                                                 URL  Year         ID  
0  https://en.wikipedia.org/wiki/The_Adventures_o...  1990  tt0098987  
1  https://en.wikipedia.org/wiki/After_Dark,_My_S...  1990  tt0098994  

[2 rows x 21 columns]

In [47]:
wiki_movies_df = wiki_recast_duration_df.copy()
wiki_movies_df.shape

(6894, 21)

### Inspect Kaggle data

In [48]:
print(kaggle_movies.info())
kaggle_movies.head(2)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 45466 entries, 0 to 45465
Data columns (total 24 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   adult                  45466 non-null  object 
 1   belongs_to_collection  4494 non-null   object 
 2   budget                 45466 non-null  object 
 3   genres                 45466 non-null  object 
 4   homepage               7782 non-null   object 
 5   id                     45466 non-null  object 
 6   imdb_id                45449 non-null  object 
 7   original_language      45455 non-null  object 
 8   original_title         45466 non-null  object 
 9   overview               44512 non-null  object 
 10  popularity             45461 non-null  object 
 11  poster_path            45080 non-null  object 
 12  production_companies   45463 non-null  object 
 13  production_countries   45463 non-null  object 
 14  release_date           45379 non-null  object 
 15  re

adult                              belongs_to_collection    budget  \
0  False  {'id': 10194, 'name': 'Toy Story Collection', ...  30000000   
1  False                                                NaN  65000000   

                                              genres  \
0  [{'id': 16, 'name': 'Animation'}, {'id': 35, '...   
1  [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...   

                               homepage    id    imdb_id original_language  \
0  http://toystory.disney.com/toy-story   862  tt0114709                en   
1                                   NaN  8844  tt0113497                en   

  original_title                                           overview  ...  \
0      Toy Story  Led by Woody, Andy's toys live happily in his ...  ...   
1        Jumanji  When siblings Judy and Peter discover an encha...  ...   

  release_date      revenue runtime  \
0   1995-10-30  373554033.0    81.0   
1   1995-12-15  262797249.0   104.0   

                                    spoken_languages    status  \
0           [{'iso_639_1': 'en', 'name': 'English'}]  Released   
1  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...  Released   

                                     tagline      title  video vote_average  \
0                                        NaN  Toy Story  False          7.7   
1  Roll the dice and unleash the excitement!    Jumanji  False          6.9   

  vote_count  
0     5415.0  
1     2413.0  

[2 rows x 24 columns]

### Duplicate rows

In [49]:
# Check for duplicated rows
print(kaggle_movies.duplicated(subset='imdb_id').sum())
kaggle_movies[kaggle_movies.duplicated(subset='imdb_id', keep=False)].sort_values('imdb_id').head()

48


adult  \
35587   Avalanche Sharks tells the story of a bikini ...   
29503   Rune Balot goes to a casino connected to the ...   
19730                                 - Written by Ørnås   
16167                                              False   
38871                                              False   

      belongs_to_collection                            budget  \
35587              2.185485  /zaSf5OG7V8X8gqFvly88zDdRm46.jpg   
29503              1.931659  /zV8bHuSL6WXoD6FWogP9j4x80bL.jpg   
19730              0.065736  /ff9qCepilowshEtG2GYWwzt2bs4.jpg   
16167                   NaN                                 0   
38871                   NaN                                 0   

                                                  genres  \
35587  [{'name': 'Odyssey Media', 'id': 17161}, {'nam...   
29503  [{'name': 'Aniplex', 'id': 2883}, {'name': 'Go...   
19730  [{'name': 'Carousel Productions', 'id': 11176}...   
16167  [{'id': 28, 'name': 'Action'}, {'id': 18, 'nam...   
38871  [{'id': 28, 'name': 'Action'}, {'id': 18, 'nam...   

                                                homepage          id  \
35587           [{'iso_3166_1': 'CA', 'name': 'Canada'}]  2014-01-01   
29503  [{'iso_3166_1': 'US', 'name': 'United States o...  2012-09-29   
19730  [{'iso_3166_1': 'CA', 'name': 'Canada'}, {'iso...  1997-08-20   
16167                                                NaN       99080   
38871                                                NaN       99080   

         imdb_id original_language                            original_title  \
35587          0              82.0  [{'iso_639_1': 'en', 'name': 'English'}]   
29503          0              68.0      [{'iso_639_1': 'ja', 'name': '日本語'}]   
19730          0             104.0  [{'iso_639_1': 'en', 'name': 'English'}]   
16167  tt0022537                en                                The Viking   
38871  tt0022537                en                                The Viking   

                                                overview  ... release_date  \
35587                                           Released  ...           22   
29503                                           Released  ...           12   
19730                                           Released  ...            1   
16167  Originally called White Thunder, American prod...  ...   1931-06-21   
38871  Originally called White Thunder, American prod...  ...   1931-06-21   

      revenue runtime                          spoken_languages    status  \
35587     NaN     NaN                                       NaN       NaN   
29503     NaN     NaN                                       NaN       NaN   
19730     NaN     NaN                                       NaN       NaN   
16167     0.0    70.0  [{'iso_639_1': 'en', 'name': 'English'}]  Released   
38871     0.0    70.0  [{'iso_639_1': 'en', 'name': 'English'}]  Released   

                                                 tagline       title  video  \
35587                                                NaN         NaN    NaN   
29503                                                NaN         NaN    NaN   
19730                                                NaN         NaN    NaN   
16167  Actually produced during the Great Newfoundlan...  The Viking  False   
38871  Actually produced during the Great Newfoundlan...  The Viking  False   

      vote_average vote_count  
35587          NaN        NaN  
29503          NaN        NaN  
19730          NaN        NaN  
16167          0.0        0.0  
38871          0.0        0.0  

[5 rows x 24 columns]

In [50]:
# Invalid IMDB IDs
kaggle_movies['imdb_id'].count() - kaggle_movies['imdb_id'].dropna().str.contains(r'tt\d{7}').sum()

3

In [51]:
# Replace IMDB ID 0 with NaN
kaggle_replace_id0_df = kaggle_movies.copy()
kaggle_replace_id0_df['imdb_id'].replace('0', np.NaN, inplace=True)
kaggle_replace_id0_df['imdb_id'].count() - kaggle_replace_id0_df['imdb_id'].dropna().str.contains(r'tt\d{7}').sum()

0

In [52]:
# Drop duplicate movies
kaggle_drop_dups_df = kaggle_replace_id0_df.drop_duplicates(subset=['imdb_id'])
kaggle_drop_dups_df.duplicated(subset='imdb_id').sum()

0

### Columns with mostly missing values

In [53]:
# Inspect columns
kaggle_drop_dups_df.isnull().mean()

adult                    0.000000
belongs_to_collection    0.901226
budget                   0.000000
genres                   0.000000
homepage                 0.828853
id                       0.000000
imdb_id                  0.000022
original_language        0.000242
original_title           0.000000
overview                 0.020961
popularity               0.000066
poster_path              0.008389
production_companies     0.000066
production_countries     0.000066
release_date             0.001850
revenue                  0.000066
runtime                  0.005681
spoken_languages         0.000066
status                   0.001850
tagline                  0.550873
title                    0.000066
video                    0.000066
vote_average             0.000066
vote_count               0.000066
dtype: float64

In [54]:
# Drop columns with more than 70% missing values
kaggle_drop_cols_df = kaggle_drop_dups_df.drop(['belongs_to_collection', 'homepage'], axis=1)
kaggle_drop_cols_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 45417 entries, 0 to 45465
Data columns (total 22 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   adult                 45417 non-null  object 
 1   budget                45417 non-null  object 
 2   genres                45417 non-null  object 
 3   id                    45417 non-null  object 
 4   imdb_id               45416 non-null  object 
 5   original_language     45406 non-null  object 
 6   original_title        45417 non-null  object 
 7   overview              44465 non-null  object 
 8   popularity            45414 non-null  object 
 9   poster_path           45036 non-null  object 
 10  production_companies  45414 non-null  object 
 11  production_countries  45414 non-null  object 
 12  release_date          45333 non-null  object 
 13  revenue               45414 non-null  float64
 14  runtime               45159 non-null  float64
 15  spoken_languages   

### Recast columns

In [55]:
# Check column values
kaggle_drop_cols_df.head(2).iloc[:, 9:]

poster_path  \
0  /rhIRbceoE9lR4veEXuwCC2wARtG.jpg   
1  /vzmL6fP7aPKNKPRTFnZmiUfciyV.jpg   

                                production_companies  \
0     [{'name': 'Pixar Animation Studios', 'id': 3}]   
1  [{'name': 'TriStar Pictures', 'id': 559}, {'na...   

                                production_countries release_date  \
0  [{'iso_3166_1': 'US', 'name': 'United States o...   1995-10-30   
1  [{'iso_3166_1': 'US', 'name': 'United States o...   1995-12-15   

       revenue  runtime                                   spoken_languages  \
0  373554033.0     81.0           [{'iso_639_1': 'en', 'name': 'English'}]   
1  262797249.0    104.0  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...   

     status                                    tagline      title  video  \
0  Released                                        NaN  Toy Story  False   
1  Released  Roll the dice and unleash the excitement!    Jumanji  False   

   vote_average  vote_count  
0           7.7      5415.0  
1           6.9      2413.0

In [56]:
# Columns to recast
cols_to_recast = {
    'adult': 'boolean',
    'budget': 'numeric',
    'id': 'numeric',
    'popularity': 'numeric',
    'release_date': 'datetime',
    'video': 'boolean'
}

### Convert adult and video to boolean type

In [57]:
# Inspect adult values
kaggle_drop_cols_df['adult'].value_counts()

False    45408
True         9
Name: adult, dtype: int64

In [58]:
# Inspect video values
kaggle_drop_cols_df['video'].value_counts()

False    45321
True        93
Name: video, dtype: int64

In [59]:
# Remove adult movies and drop both columns
kaggle_drop_bool_df = kaggle_drop_cols_df.query('adult != "True"').drop(['adult', 'video'], axis=1).copy()
kaggle_drop_bool_df.shape

(45408, 20)

### Convert release date to datetime type

In [60]:
# Recast release date to datetime
kaggle_recast_date_df = kaggle_drop_bool_df.copy()
kaggle_recast_date_df['release_date'] = pd.to_datetime(kaggle_recast_date_df['release_date'])
kaggle_recast_date_df['release_date']

0       1995-10-30
1       1995-12-15
2       1995-12-22
3       1995-12-22
4       1995-02-10
           ...    
45461          NaT
45462   2011-11-17
45463   2003-08-01
45464   1917-10-21
45465   2017-06-09
Name: release_date, Length: 45408, dtype: datetime64[ns]

### Convert budget, ID, and popularity to numeric

In [61]:
# Recast budget and ID to integer
kaggle_recast_numeric_df = kaggle_recast_date_df.copy()
kaggle_recast_numeric_df['budget'] = kaggle_recast_numeric_df['budget'].astype(int)
kaggle_recast_numeric_df['id'] = kaggle_recast_numeric_df['id'].astype(int)

# Recast popularity to float
kaggle_recast_numeric_df['popularity'] = kaggle_recast_numeric_df['popularity'].astype(float)
kaggle_recast_numeric_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 45408 entries, 0 to 45465
Data columns (total 20 columns):
 #   Column                Non-Null Count  Dtype         
---  ------                --------------  -----         
 0   budget                45408 non-null  int64         
 1   genres                45408 non-null  object        
 2   id                    45408 non-null  int64         
 3   imdb_id               45407 non-null  object        
 4   original_language     45397 non-null  object        
 5   original_title        45408 non-null  object        
 6   overview              44456 non-null  object        
 7   popularity            45405 non-null  float64       
 8   poster_path           45027 non-null  object        
 9   production_companies  45405 non-null  object        
 10  production_countries  45405 non-null  object        
 11  release_date          45325 non-null  datetime64[ns]
 12  revenue               45405 non-null  float64       
 13  runtime         

In [62]:
kaggle_movies_df = kaggle_recast_numeric_df.copy()
kaggle_movies_df.shape

(45408, 20)

### Inspect ratings data

In [63]:
print(kaggle_ratings.info(null_counts=True))
kaggle_ratings.head(2)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 26024289 entries, 0 to 26024288
Data columns (total 4 columns):
 #   Column     Non-Null Count     Dtype  
---  ------     --------------     -----  
 0   userId     26024289 non-null  int64  
 1   movieId    26024289 non-null  int64  
 2   rating     26024289 non-null  float64
 3   timestamp  26024289 non-null  int64  
dtypes: float64(1), int64(3)
memory usage: 794.2 MB
None


userId  movieId  rating   timestamp
0       1      110     1.0  1425941529
1       1      147     4.5  1425942435

### Save data

In [66]:
wiki_movies_df.to_pickle(data_path + 'wiki_movies.pkl')
pd.read_pickle(data_path + 'wiki_movies.pkl').head(2)

Based on Release date  \
0                        [Characters, by Rex Weiner]   1990-07-11   
1  [the novel, After Dark, My Sweet, by, Jim Thom...   1990-05-17   

         Country      Budget  Box office Cinematographer(s)      Director  \
0  United States  20000000.0  21400000.0        Oliver Wood  Renny Harlin   
1  United States   6000000.0   2700000.0       Mark Plummer   James Foley   

        Distributor        Editor(s) Language(s)  ...  \
0  20th Century Fox  Michael Tronick     English  ...   
1   Avenue Pictures  Howard E. Smith     English  ...   

                   Producer(s) Production company(s) Duration  \
0   [Steve Perry, Joel Silver]       Silver Pictures    102.0   
1  [Ric Kidney, Robert Redlin]       Avenue Pictures    114.0   

                      Writer(s)  \
0   [David Arnott, James Cappe]   
1  [James Foley, Robert Redlin]   

                                               Stars  \
0  [Andrew Dice Clay, Wayne Newton, Priscilla Pre...   
1  [Jason Patric, Rachel Ward, Bruce Dern, George...   

                               IMDB link                            Title  \
0  https://www.imdb.com/title/tt0098987/  The Adventures of Ford Fairlane   
1  https://www.imdb.com/title/tt0098994/             After Dark, My Sweet   

                                                 URL  Year         ID  
0  https://en.wikipedia.org/wiki/The_Adventures_o...  1990  tt0098987  
1  https://en.wikipedia.org/wiki/After_Dark,_My_S...  1990  tt0098994  

[2 rows x 21 columns]

In [67]:
kaggle_movies_df.to_pickle(data_path + 'kaggle_movies.pkl')
pd.read_pickle(data_path + 'kaggle_movies.pkl').head(2)

budget                                             genres    id  \
0  30000000  [{'id': 16, 'name': 'Animation'}, {'id': 35, '...   862   
1  65000000  [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...  8844   

     imdb_id original_language original_title  \
0  tt0114709                en      Toy Story   
1  tt0113497                en        Jumanji   

                                            overview  popularity  \
0  Led by Woody, Andy's toys live happily in his ...   21.946943   
1  When siblings Judy and Peter discover an encha...   17.015539   

                        poster_path  \
0  /rhIRbceoE9lR4veEXuwCC2wARtG.jpg   
1  /vzmL6fP7aPKNKPRTFnZmiUfciyV.jpg   

                                production_companies  \
0     [{'name': 'Pixar Animation Studios', 'id': 3}]   
1  [{'name': 'TriStar Pictures', 'id': 559}, {'na...   

                                production_countries release_date  \
0  [{'iso_3166_1': 'US', 'name': 'United States o...   1995-10-30   
1  [{'iso_3166_1': 'US', 'name': 'United States o...   1995-12-15   

       revenue  runtime                                   spoken_languages  \
0  373554033.0     81.0           [{'iso_639_1': 'en', 'name': 'English'}]   
1  262797249.0    104.0  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...   

     status                                    tagline      title  \
0  Released                                        NaN  Toy Story   
1  Released  Roll the dice and unleash the excitement!    Jumanji   

   vote_average  vote_count  
0           7.7      5415.0  
1           6.9      2413.0